File: core/bases/base_service.py
from typing import Any, Dict, Generic, List, Type, TypeVar, Union
from pydantic import BaseModel
from sqlmodel import SQLModel

from src.core.bases.base_repository import BaseRepository
from src.core import exceptions
from src.core.response.schemas import PaginatedResponse
from src.core.schemas.fields import DynamicFormConfig, ModelDefinition
from src.core.services.field_service import FieldService

T = TypeVar("T")
CreateSchemaType = TypeVar("CreateSchemaType", bound=BaseModel)
UpdateSchemaType = TypeVar("UpdateSchemaType", bound=BaseModel)


class BaseService(Generic[T]):
    """Base service class with common CRUD operations and standardized responses."""

    _instance = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self, repository: BaseRepository) -> None:
        self.repository = repository

    async def _return_one_data(self, data: Any) -> Any:
        return data

    async def _return_multi_data(self, data: List) -> Any:
        return [await self._return_one_data(item) for item in data]

    async def get_all(self, include_deleted: bool = False, **filters) -> Dict[str, Any]:
        """Get all items without pagination."""
        try:
            items = await self.repository.get_all(
                include_deleted=include_deleted, **filters
            )
            return {
                "data": await self._return_multi_data(items),
                "message": f"Retrieved {len(items)} items successfully",
            }
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error retrieving all items: {str(e)}"
            ) from e

    async def get_by_id(
        self, item_id: Any, include_deleted: bool = False, **filters
    ) -> Dict[str, Any]:
        """Get a single item by ID."""
        try:
            item = await self.repository.get(
                item_id=item_id, include_deleted=include_deleted, **filters
            )

            if not item:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found"
                )

            return {
                "data": await self._return_one_data(item),
                "message": "Item retrieved successfully",
            }
        except exceptions.NotFoundException:
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error retrieving item: {str(e)}"
            ) from e

    async def get_list(
        self,
        page: int = 1,
        per_page: int = 10,
        include_deleted: bool = False,
        **filters,
    ) -> Dict[str, Any]:
        """Get paginated list of items."""
        try:
            # Validate pagination parameters
            if page < 1:
                page = 1
            if per_page < 1 or per_page > 100:
                per_page = 10

            result = await self.repository.list(
                page=page, per_page=per_page, include_deleted=include_deleted, **filters
            )

            return await self._result_to_pagenation_response(result)
        except Exception as e:
            raise e
            raise exceptions.ServiceException(
                detail=f"Error retrieving items list: {str(e)}"
            ) from e

    async def get_many(
        self, skip: int = 0, limit: int = 100, include_deleted: bool = False, **filters
    ) -> Dict[str, Any]:
        """Get multiple items without pagination metadata."""
        try:
            items = await self.repository.get_many(
                skip=skip, limit=limit, include_deleted=include_deleted, **filters
            )

            return {
                "data": await self._return_multi_data(items),
                "message": f"Retrieved {len(items)} items successfully",
            }
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error retrieving items: {str(e)}"
            ) from e

    async def create(
        self,
        obj_in: Any,
        **additional_data,
    ) -> Dict[str, Any]:
        """Create a new item."""
        try:
            # Convert Pydantic model to dict if needed
            if isinstance(obj_in, BaseModel):
                create_data = obj_in.model_dump(exclude_unset=True)
            else:
                create_data = obj_in.copy()

            # Merge with additional data
            create_data.update(additional_data)

            # Validate business rules before creation
            await self._validate_create(create_data)
            item = await self.repository.create(create_data)

            return {
                "data": await self._return_one_data(item),
                "message": "Item created successfully",
            }
        except exceptions.ValidationException:
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error creating item: {str(e)}"
            ) from e

    async def update(
        self,
        item_id: Any,
        obj_in: Union[Dict[str, Any], BaseModel],
        exclude_unset: bool = True,
        **additional_data,
    ) -> Dict[str, Any]:
        """Update an existing item."""
        try:
            # Check if item exists
            existing_item = await self.repository.get(item_id)
            if not existing_item:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found"
                )

            # Convert Pydantic model to dict if needed
            if isinstance(obj_in, BaseModel):
                update_data = obj_in.model_dump(exclude_unset=exclude_unset)
            else:
                update_data = obj_in.copy()

            # Merge with additional data
            update_data.update(additional_data)

            # Validate business rules before update
            await self._validate_update(item_id, update_data, existing_item)

            updated_item = await self.repository.update(
                item_id=item_id, obj_in=update_data
            )

            if not updated_item:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found during update"
                )
            return {
                "data": await self._return_one_data(updated_item),
                "message": "Item updated successfully",
            }
        except (exceptions.NotFoundException, exceptions.ValidationException) as e:
            raise e
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error updating item: {str(e)}"
            ) from e

    async def soft_delete(self, item_id: int) -> Dict[str, Any]:
        """Soft delete an item."""
        try:
            # Check if item exists
            existing_item = await self.repository.get(item_id)

            if not existing_item:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found"
                )

            # Validate if soft delete is allowed
            await self._validate_delete(item_id, existing_item)

            success = await self.repository.soft_delete(item_id)

            if not success:
                raise exceptions.OperationException(
                    detail=f"Failed to soft delete item with id {item_id}"
                )

            return {"data": None, "message": "Item soft deleted successfully"}
        except (exceptions.NotFoundException, exceptions.ValidationException):
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error soft deleting item: {str(e)}"
            ) from e

    async def restore(self, item_id: Any) -> Dict[str, Any]:
        """Restore a soft deleted item."""
        try:
            success = await self.repository.restore(item_id)

            if not success:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found or not deleted"
                )

            return {"data": None, "message": "Item restored successfully"}
        except exceptions.NotFoundException:
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error restoring item: {str(e)}"
            ) from e

    async def force_delete(self, item_id: Any) -> Dict[str, Any]:
        """Permanently delete an item."""
        try:
            # Check if item exists
            existing_item = await self.repository.get(item_id, include_deleted=True)
            if not existing_item:
                raise exceptions.NotFoundException(
                    detail=f"Item with id {item_id} not found"
                )

            # Validate if force delete is allowed
            await self._validate_force_delete(item_id, existing_item)

            success = await self.repository.force_delete(item_id)

            if not success:
                raise exceptions.OperationException(
                    detail=f"Failed to delete item with id {item_id}"
                )

            return {"data": None, "message": "Item permanently deleted successfully"}
        except (exceptions.NotFoundException, exceptions.ValidationException):
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error deleting item: {str(e)}"
            ) from e

    async def exists(
        self, item_id: Any, include_deleted: bool = False
    ) -> Dict[str, Any]:
        """Check if an item exists."""
        try:
            exists = await self.repository.exists(
                item_id, include_deleted=include_deleted
            )

            return {
                "data": {"exists": exists},
                "message": f"Item {'exists' if exists else 'does not exist'}",
            }
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error checking item existence: {str(e)}"
            ) from e

    async def count(self, include_deleted: bool = False, **filters) -> Dict[str, Any]:
        """Count items matching filters."""
        try:
            count = await self.repository.count(
                include_deleted=include_deleted, **filters
            )

            return {"data": {"count": count}, "message": f"Found {count} items"}
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error counting items: {str(e)}"
            ) from e

    async def bulk_create(
        self, objs_in: List[Union[Dict[str, Any], BaseModel]], **additional_data
    ) -> Dict[str, Any]:
        """Create multiple items in bulk."""
        try:
            create_data_list = []
            for obj_in in objs_in:
                if isinstance(obj_in, BaseModel):
                    create_data = obj_in.model_dump(exclude_unset=True)
                else:
                    create_data = obj_in.copy()

                await self._validate_create(create_data)

                create_data.update(additional_data)
                create_data_list.append(create_data)

            items = await self.repository.bulk_create(create_data_list)

            return {
                "data": await self._return_multi_data(items),
                "message": f"Successfully created {len(items)} items",
            }
        except exceptions.ValidationException:
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error in bulk creating items: {str(e)}"
            ) from e

    async def bulk_update(
        self, objs_in: List[Union[Dict[str, Any], BaseModel]], **additional_data
    ) -> Dict[str, Any]:
        """Create multiple items in bulk."""
        try:
            update_data_list = []
            for obj_in in objs_in:
                id = (
                    obj_in.get("id")
                    if isinstance(obj_in, dict)
                    else obj_in.id  # type:ignore
                )
                if not id:
                    raise exceptions.ValidationException(
                        detail="ID is required for bulk update"
                    )
                obj = (
                    obj_in.get("data")
                    if isinstance(obj_in, dict)
                    else obj_in.data  # type:ignore
                )
                if isinstance(obj, BaseModel):
                    update_data = obj.model_dump(exclude_unset=True)
                else:
                    update_data = obj.copy()  # type:ignore

                await self._validate_update(update_data.get("id"), update_data, await self.repository.get(id))  # type: ignore

                update_data.update(additional_data)
                update_data_list.append(update_data)

            items = await self.repository.bulk_update(update_data_list)

            return {
                "data": await self._return_multi_data(items),
                "message": f"Successfully updated {len(items)} items",
            }
        except exceptions.ValidationException:
            raise
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error in bulk creating items: {str(e)}"
            ) from e

    async def search(self, query: str):
        """Search items based on query string."""
        try:
            if not self.repository._search_fields:
                raise exceptions.ServiceException(
                    detail="Search fields not defined in repository"
                )
            result = await self.repository.search(query, self.repository._search_fields)
            return await self._result_to_pagenation_response(result)
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error searching items: {str(e)}"
            ) from e

    async def bulk_delete(self, ids: List[int]):
        try:
            result = await self.repository.bulk_delete(ids)
            return {
                "data": result,
                "message": f"Successfully deleted {result} items",
            }
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error searching items: {str(e)}"
            ) from e

    async def get_logs(self, **filters):
        try:
            result = await self.repository.get_logs(**filters)
            return {
                "items": result.data,
                "total": result.total,
                "page": result.page,
                "per_page": result.per_page,
                "pages": result.pages,
                "message": result.message,
            }
        except Exception as e:
            raise exceptions.ServiceException(
                detail=f"Error retrieving logs: {str(e)}"
            ) from e

    # Validation methods to be overridden by subclasses
    async def _validate_create(self, create_data: Dict[str, Any]) -> None:
        """Validate data before creation. Override in subclasses."""
        pass

    async def _validate_update(
        self, item_id: Any, update_data: Dict[str, Any], existing_item: T
    ) -> None:
        """Validate data before update. Override in subclasses."""
        pass

    async def _validate_delete(self, item_id: Any, existing_item: T) -> None:
        """Validate before soft delete. Override in subclasses."""
        pass

    async def _validate_force_delete(self, item_id: Any, existing_item: T) -> None:
        """Validate before force delete. Override in subclasses."""
        pass

    async def _result_to_pagenation_response(self, result: PaginatedResponse) -> dict:
        return {
            "items": await self._return_multi_data(result.data),
            "total": result.total,
            "page": result.page,
            "per_page": result.per_page,
            "pages": result.pages,
            "message": result.message,
        }

    def get_model_field_definitions(
        self, model_class: Type[SQLModel]
    ) -> ModelDefinition:
        """Get model field definitions."""
        return FieldService.get_model_definition(model_class)

    def get_dynamic_form_config(self, model_class: Type[SQLModel]) -> DynamicFormConfig:
        """Get dynamic form configuration."""
        return FieldService.get_dynamic_form_config(model_class)

    def get_enum_definitions(self, enum_class: str, app_name: str) -> Dict[str, Any]:
        """Get enum definitions."""
        try:
            import importlib
            import enum as py_enum

            path = f"src.apps.{app_name}.utils.enums"
            enums_module = importlib.import_module(path)
            enum_cls = getattr(enums_module, enum_class, None)
            print("Importing enum class", enum_cls)
            if enum_cls:
                if isinstance(enum_cls, type) and issubclass(enum_cls, py_enum.Enum):
                    values = [member.value for member in enum_cls]
                else:
                    try:
                        values = list(enum_cls)  # type:ignore
                    except TypeError:
                        values = [enum_cls]
                return {
                    "data": values,
                    "message": f"Enum definitions for {enum_class} retrieved successfully",
                }
            return {
                "data": [],
                "message": f"Enum definitions for {enum_class} retrieved successfully",
            }
        except Exception as e:
            print(e)
            raise exceptions.ServiceException(
                detail=f"Error retrieving enum definitions: {str(e)}"
            ) from e

File: RootCore/Bases/BaseService.cs
using RootCore.Database;
using RootCore.Exceptions;
using RootCore.Response;
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace RootCore.Bases
{
    /// <summary>
    /// Base service class with common CRUD operations and standardized responses.
    /// </summary>
    /// <typeparam name="T">Entity type that inherits from BaseModel</typeparam>
    public class BaseService<T> where T : BaseModel
    {
        protected readonly BaseRepository<T> _repository;

        public BaseService(BaseRepository<T> repository)
        {
            _repository = repository;
        }

        /// <summary>
        /// Transform data before returning (override in subclasses).
        /// </summary>
        protected virtual Task<T> TransformOneDataAsync(T data)
        {
            return Task.FromResult(data);
        }

        /// <summary>
        /// Transform multiple data items before returning.
        /// </summary>
        protected virtual async Task<List<T>> TransformMultiDataAsync(List<T> data)
        {
            var result = new List<T>();
            foreach (var item in data)
            {
                result.Add(await TransformOneDataAsync(item));
            }
            return result;
        }

        /// <summary>
        /// Get all items without pagination.
        /// </summary>
        public virtual async Task<ApiResponse<List<T>>> GetAllAsync(bool includeDeleted = false)
        {
            try
            {
                var items = await _repository.GetAllAsync(includeDeleted);
                var transformedItems = await TransformMultiDataAsync(items);
                return new ApiResponse<List<T>>(
                    transformedItems, 
                    $"Retrieved {items.Count} items successfully"
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error retrieving all items: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get a single item by ID.
        /// </summary>
        public virtual async Task<ApiResponse<T>> GetByIdAsync(int id, bool includeDeleted = false)
        {
            try
            {
                var item = await _repository.GetByIdAsync(id, includeDeleted);
                
                if (item == null)
                {
                    throw new NotFoundException($"Item with id {id} not found");
                }

                var transformedItem = await TransformOneDataAsync(item);
                return new ApiResponse<T>(transformedItem, "Item retrieved successfully");
            }
            catch (NotFoundException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error retrieving item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get paginated list of items.
        /// </summary>
        public virtual async Task<PaginatedResponse<T>> GetListAsync(
            int page = 1,
            int perPage = 10,
            bool includeDeleted = false,
            Expression<Func<T, bool>>? filter = null)
        {
            try
            {
                // Validate pagination parameters
                if (page < 1) page = 1;
                if (perPage < 1 || perPage > 100) perPage = 10;

                var result = await _repository.GetListAsync(page, perPage, includeDeleted, filter);
                var transformedItems = await TransformMultiDataAsync(result.Items);
                
                return new PaginatedResponse<T>(
                    transformedItems,
                    result.Total,
                    result.Page,
                    result.PerPage,
                    result.Message
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error retrieving items list: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Get multiple items without pagination metadata.
        /// </summary>
        public virtual async Task<ApiResponse<List<T>>> GetManyAsync(
            int skip = 0,
            int limit = 100,
            bool includeDeleted = false,
            Expression<Func<T, bool>>? filter = null)
        {
            try
            {
                var items = await _repository.GetManyAsync(skip, limit, includeDeleted, filter);
                var transformedItems = await TransformMultiDataAsync(items);
                
                return new ApiResponse<List<T>>(
                    transformedItems,
                    $"Retrieved {items.Count} items successfully"
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error retrieving items: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Create a new item.
        /// </summary>
        public virtual async Task<ApiResponse<T>> CreateAsync(T entity)
        {
            try
            {
                // Validate business rules before creation
                await ValidateCreateAsync(entity);
                
                var item = await _repository.CreateAsync(entity);
                var transformedItem = await TransformOneDataAsync(item);
                
                return new ApiResponse<T>(transformedItem, "Item created successfully");
            }
            catch (ValidationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error creating item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Update an existing item.
        /// </summary>
        public virtual async Task<ApiResponse<T>> UpdateAsync(int id, T entity)
        {
            try
            {
                // Check if item exists
                var existingItem = await _repository.GetByIdAsync(id);
                if (existingItem == null)
                {
                    throw new NotFoundException($"Item with id {id} not found");
                }

                // Validate business rules before update
                await ValidateUpdateAsync(id, entity, existingItem);

                entity.Id = id;
                var updatedItem = await _repository.UpdateAsync(entity);
                var transformedItem = await TransformOneDataAsync(updatedItem);
                
                return new ApiResponse<T>(transformedItem, "Item updated successfully");
            }
            catch (NotFoundException)
            {
                throw;
            }
            catch (ValidationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error updating item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Soft delete an item.
        /// </summary>
        public virtual async Task<ApiResponse<object>> SoftDeleteAsync(int id)
        {
            try
            {
                // Check if item exists
                var existingItem = await _repository.GetByIdAsync(id);
                if (existingItem == null)
                {
                    throw new NotFoundException($"Item with id {id} not found");
                }

                // Validate if soft delete is allowed
                await ValidateDeleteAsync(id, existingItem);

                var success = await _repository.SoftDeleteAsync(id);
                
                if (!success)
                {
                    throw new OperationException($"Failed to soft delete item with id {id}");
                }

                return new ApiResponse<object>(null!, "Item soft deleted successfully");
            }
            catch (NotFoundException)
            {
                throw;
            }
            catch (ValidationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error soft deleting item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Restore a soft deleted item.
        /// </summary>
        public virtual async Task<ApiResponse<object>> RestoreAsync(int id)
        {
            try
            {
                var success = await _repository.RestoreAsync(id);
                
                if (!success)
                {
                    throw new NotFoundException($"Item with id {id} not found or not deleted");
                }

                return new ApiResponse<object>(null!, "Item restored successfully");
            }
            catch (NotFoundException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error restoring item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Permanently delete an item.
        /// </summary>
        public virtual async Task<ApiResponse<object>> ForceDeleteAsync(int id)
        {
            try
            {
                // Check if item exists
                var existingItem = await _repository.GetByIdAsync(id, includeDeleted: true);
                if (existingItem == null)
                {
                    throw new NotFoundException($"Item with id {id} not found");
                }

                // Validate if force delete is allowed
                await ValidateForceDeleteAsync(id, existingItem);

                var success = await _repository.ForceDeleteAsync(id);
                
                if (!success)
                {
                    throw new OperationException($"Failed to delete item with id {id}");
                }

                return new ApiResponse<object>(null!, "Item permanently deleted successfully");
            }
            catch (NotFoundException)
            {
                throw;
            }
            catch (ValidationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error deleting item: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Check if an item exists.
        /// </summary>
        public virtual async Task<ApiResponse<Dictionary<string, bool>>> ExistsAsync(
            int id, 
            bool includeDeleted = false)
        {
            try
            {
                var exists = await _repository.ExistsAsync(id, includeDeleted);
                
                return new ApiResponse<Dictionary<string, bool>>(
                    new Dictionary<string, bool> { { "exists", exists } },
                    $"Item {(exists ? "exists" : "does not exist")}"
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error checking item existence: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Count items matching filters.
        /// </summary>
        public virtual async Task<ApiResponse<Dictionary<string, int>>> CountAsync(
            bool includeDeleted = false,
            Expression<Func<T, bool>>? filter = null)
        {
            try
            {
                var count = await _repository.CountAsync(includeDeleted, filter);
                
                return new ApiResponse<Dictionary<string, int>>(
                    new Dictionary<string, int> { { "count", count } },
                    $"Found {count} items"
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error counting items: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Create multiple items in bulk.
        /// </summary>
        public virtual async Task<ApiResponse<List<T>>> BulkCreateAsync(List<T> entities)
        {
            try
            {
                foreach (var entity in entities)
                {
                    await ValidateCreateAsync(entity);
                }

                var items = await _repository.BulkCreateAsync(entities);
                var transformedItems = await TransformMultiDataAsync(items);
                
                return new ApiResponse<List<T>>(
                    transformedItems,
                    $"Successfully created {items.Count} items"
                );
            }
            catch (ValidationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error in bulk creating items: {ex.Message}", ex);
            }
        }

        /// <summary>
        /// Bulk delete multiple items.
        /// </summary>
        public virtual async Task<ApiResponse<int>> BulkDeleteAsync(List<int> ids)
        {
            try
            {
                var count = await _repository.BulkDeleteAsync(ids);
                
                return new ApiResponse<int>(
                    count,
                    $"Successfully deleted {count} items"
                );
            }
            catch (Exception ex)
            {
                throw new ServiceException($"Error bulk deleting items: {ex.Message}", ex);
            }
        }

        // Validation methods to be overridden by subclasses
        protected virtual Task ValidateCreateAsync(T entity)
        {
            return Task.CompletedTask;
        }

        protected virtual Task ValidateUpdateAsync(int id, T entity, T existingItem)
        {
            return Task.CompletedTask;
        }

        protected virtual Task ValidateDeleteAsync(int id, T existingItem)
        {
            return Task.CompletedTask;
        }

        protected virtual Task ValidateForceDeleteAsync(int id, T existingItem)
        {
            return Task.CompletedTask;
        }
    }
}
